File: app/src/main/java/net/bloople/stories/IndexingActivity.java
package net.bloople.stories;

import android.Manifest;
import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.pm.PackageManager;
import android.os.AsyncTask;
import android.os.Bundle;
import android.os.Environment;
import android.support.annotation.NonNull;
import android.view.KeyEvent;
import android.view.View;
import android.view.inputmethod.EditorInfo;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ProgressBar;
import android.widget.TextView;
import android.widget.Toast;

import java.io.File;
import java.io.IOException;
import java.util.ArrayList;

public class IndexingActivity extends Activity implements Indexable {
    // Storage Permissions
    private static final int REQUEST_EXTERNAL_STORAGE = 1;
    private static String[] PERMISSIONS_STORAGE = {
            Manifest.permission.READ_EXTERNAL_STORAGE,
            Manifest.permission.WRITE_EXTERNAL_STORAGE
    };

    private ProgressBar progressBar;
    private Button indexButton;
    private String indexRoot;
    private boolean canAccessFiles;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_indexing);

        if(android.os.Build.VERSION.SDK_INT >= 23) {
            int permission = checkSelfPermission(Manifest.permission.WRITE_EXTERNAL_STORAGE);

            canAccessFiles = false;
            if(permission == PackageManager.PERMISSION_GRANTED) canAccessFiles = true;
            else requestPermissions(PERMISSIONS_STORAGE, REQUEST_EXTERNAL_STORAGE);
        }
        else {
            canAccessFiles = true;
        }

        progressBar = (ProgressBar)findViewById(R.id.indexing_progress);

        indexButton = (Button)findViewById(R.id.index_button);
        indexButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                indexButton.setEnabled(false);

                if(canAccessFiles) {
                    IndexingTask indexer = new IndexingTask(IndexingActivity.this,
                            IndexingActivity.this);
                    indexer.execute(indexRoot);
                }
            }
        });

        Button deleteIndexButton = (Button)findViewById(R.id.delete_index_button);
        deleteIndexButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                DatabaseHelper.deleteDatabase(IndexingActivity.this);
                Toast.makeText(IndexingActivity.this, "Index deleted.", Toast.LENGTH_SHORT).show();
            }
        });

        loadPreferences();

        final EditText indexDirectoryText = (EditText)findViewById(R.id.index_directory);
        indexDirectoryText.setText(indexRoot);

        indexDirectoryText.setOnEditorActionListener(new TextView.OnEditorActionListener() {
            public boolean onEditorAction(TextView v, int actionId, KeyEvent event) {
                if((event != null && (event.getKeyCode() == KeyEvent.KEYCODE_ENTER)) ||
                        (actionId == EditorInfo.IME_ACTION_DONE)) {
                    indexRoot = indexDirectoryText.getText().toString();
                    savePreferences();
                }
                return false;
            }
        });
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);

        if(resultCode == RESULT_CANCELED) finish();
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String permissions[],
                                           @NonNull int[] grantResults) {
        if(requestCode == REQUEST_EXTERNAL_STORAGE && grantResults.length > 0 &&
                grantResults[0] == PackageManager.PERMISSION_GRANTED) canAccessFiles = true;
    }

    private SharedPreferences preferences() {
        return getApplicationContext().getSharedPreferences("main", Context.MODE_PRIVATE);
    }

    private void loadPreferences() {
        SharedPreferences preferences = preferences();

        indexRoot = preferences.getString("index-root",
                Environment.getExternalStorageDirectory().getAbsolutePath());
    }

    private void savePreferences() {
        SharedPreferences.Editor editor = preferences().edit();
        editor.putString("index-root", indexRoot);
        editor.apply();
    }

    public void onIndexingProgress(int progress, int max) {
        progressBar.setProgress(progress);
        progressBar.setMax(max);
    }

    public void onIndexingComplete(int count) {
        indexButton.setEnabled(true);
        Toast.makeText(IndexingActivity.this, "Indexing complete, " + count + " stories indexed.",
                Toast.LENGTH_LONG).show();
    }
}

File: app/src/main/java/net/bloople/stories/IndexingActivity.kt
package net.bloople.stories

import android.Manifest
import net.bloople.stories.DatabaseHelper.deleteDatabase
import android.app.Activity
import android.os.Bundle
import android.content.pm.PackageManager
import android.view.inputmethod.EditorInfo
import android.content.Intent
import android.content.SharedPreferences
import android.os.Environment
import android.view.KeyEvent
import android.widget.Button
import android.widget.EditText
import android.widget.ProgressBar
import android.widget.Toast

class IndexingActivity : Activity(), Indexable {
    private lateinit var progressBar: ProgressBar
    private lateinit var indexButton: Button
    private lateinit var indexRoot: String
    private var canAccessFiles = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_indexing)

        val permission = checkSelfPermission(Manifest.permission.WRITE_EXTERNAL_STORAGE)
        if(permission == PackageManager.PERMISSION_GRANTED) canAccessFiles = true
        else requestPermissions(PERMISSIONS_STORAGE, REQUEST_EXTERNAL_STORAGE)

        progressBar = findViewById(R.id.indexing_progress)
        indexButton = findViewById(R.id.index_button)
        indexButton.setOnClickListener {
            indexButton.isEnabled = false
            if(canAccessFiles) {
                val indexer = IndexingTask(this@IndexingActivity,this@IndexingActivity)
                indexer.execute(indexRoot)
            }
        }

        val deleteIndexButton: Button = findViewById(R.id.delete_index_button)
        deleteIndexButton.setOnClickListener {
            deleteDatabase(this@IndexingActivity)
            Toast.makeText(this@IndexingActivity, "Index deleted.", Toast.LENGTH_SHORT).show()
        }

        loadPreferences()

        val indexDirectoryText: EditText = findViewById(R.id.index_directory)
        indexDirectoryText.setText(indexRoot)
        indexDirectoryText.setOnEditorActionListener { _, actionId, event ->
            if(event != null && event.keyCode == KeyEvent.KEYCODE_ENTER || actionId == EditorInfo.IME_ACTION_DONE) {
                indexRoot = indexDirectoryText.text.toString()
                savePreferences()
            }
            false
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent) {
        super.onActivityResult(requestCode, resultCode, data)
        if(resultCode == RESULT_CANCELED) finish()
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        if(requestCode == REQUEST_EXTERNAL_STORAGE && grantResults.isNotEmpty()
            && grantResults[0] == PackageManager.PERMISSION_GRANTED) canAccessFiles = true
    }

    private fun preferences(): SharedPreferences {
        return applicationContext.getSharedPreferences("main", MODE_PRIVATE)
    }

    private fun loadPreferences() {
        val preferences = preferences()
        indexRoot = preferences.getString("index-root", Environment.getExternalStorageDirectory().absolutePath).toString()
    }

    private fun savePreferences() {
        val editor = preferences().edit()
        editor.putString("index-root", indexRoot)
        editor.apply()
    }

    override fun onIndexingProgress(progress: Int, max: Int) {
        progressBar.progress = progress
        progressBar.max = max
    }

    override fun onIndexingComplete(count: Int) {
        setResult(RESULT_OK)
        indexButton.isEnabled = true
        Toast.makeText(
            this@IndexingActivity, "Indexing complete, $count stories indexed.",
            Toast.LENGTH_LONG
        ).show()
    }

    companion object {
        // Storage Permissions
        private const val REQUEST_EXTERNAL_STORAGE = 1
        private val PERMISSIONS_STORAGE = arrayOf(
            Manifest.permission.READ_EXTERNAL_STORAGE,
            Manifest.permission.WRITE_EXTERNAL_STORAGE
        )
    }
}